pkg/oci-handler/oci.go

```diff
@@ -0,0 +1,324 @@
+			if err != nil {
+				logger.Errorf("instantiating operator %q: %v", op.Name(), err)
+			}
+			if opInst == nil {
+				logger.Debugf("> skipped %s", op.Name())
+				continue
+			}
+			instance.imageOperatorInstances = append(instance.imageOperatorInstances, opInst)
+		}
+	}
+
+	for _, op := range operators.GetDataOperators() {
+		logger.Debugf("found data op %q", op.Name())
+
+		// Lazily initialize operator
+		// TODO: global params should be filled out from a config file or such; maybe it's a better idea not to
+		// lazily initialize operators at all, but just hand over the config. The "lazy" stuff could then be done
+		// if the operator is instantiated and needs to do work
+		err := op.Init(op.GlobalParamDescs().ToParams())
+		if err != nil {
+			return nil, fmt.Errorf("initializing operator %q: %w", op.Name(), err)
+		}
+
+		opInst, err := op.InstantiateDataOperator(gadgetCtx)
+		if err != nil {
+			logger.Errorf("instantiating operator %q: %v", op.Name(), err)
+		}
+		if opInst == nil {
+			logger.Debugf("> skipped %s", op.Name())
+			continue
+		}
+		instance.dataOperatorInstances = append(instance.dataOperatorInstances, opInst)
+	}
+
+	// Sort dataOperators based on their priority
+	sort.Slice(instance.dataOperatorInstances, func(i, j int) bool {
+		return instance.dataOperatorInstances[i].Priority() < instance.dataOperatorInstances[j].Priority()
+	})
+
+	return instance, nil
+}
+
+func (o *OciHandlerInstance) ParamDescs() params.ParamDescs {
+	return nil
+}
+
+func (o *OciHandlerInstance) Prepare() error {
+	gadgetParams := make([]*api.Param, 0)
+	for _, opInst := range o.imageOperatorInstances {
+		err := opInst.Prepare(o.gadgetCtx)
+		if err != nil {
+			o.gadgetCtx.Logger().Errorf("preparing operator %q: %v", opInst.Name(), err)
+		}
+
+		// Add gadget params
+		for _, p := range opInst.GadgetParams() {
+			p.Prefix = "gadget."
+		}
+
+		// After calling Params(), we can extract GadgetParams, so let's add them
+		gadgetParams = append(gadgetParams, opInst.GadgetParams()...)
+	}
+
+	for _, opInst := range o.dataOperatorInstances {
+		opParamPrefix := fmt.Sprintf("operator.%s.", opInst.Name())
+
+		// Get and fill params
+		params := opInst.ParamDescs(o.gadgetCtx).ToParams()
+
+		err := params.CopyFromMap(o.paramValueMap, opParamPrefix)
+		if err != nil {
+			return fmt.Errorf("validating params for operator %q: %w", opInst.Name(), err)
+		}
+
+		err = opInst.Prepare(o.gadgetCtx, params)
+		if err != nil {
+			o.gadgetCtx.Logger().Errorf("preparing operator %q: %v", opInst.Name(), err)
+			continue
+		}
+
+		// Second pass params; this time the operator had the chance to prepare itself based on DataSources, etc.
+		// this mainly is postponed to read default values that might differ from before; this second pass is
+		// what is handed over to the remote end
+		pd := opInst.ParamDescs(o.gadgetCtx)
+		gadgetParams = append(gadgetParams, api.ParamDescsToParams(pd, opParamPrefix)...)
```

Comment:
I think I understand what you're trying to do: Get params after having prepared the operator instance because the params might differ (from what is returned in line 259) because of the DataSources. Is it correct? If so, I don't see any example of a data operator with such logic. Am I missing something?

Reply:
_Default columns_ like mentioned above would be an example I'm currently working on.